x-pack/test/api_integration/apis/ml/datafeeds/update.ts

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import expect from '@kbn/expect';
+import { FtrProviderContext } from '../../../ftr_provider_context';
+import { USER } from '../../../../functional/services/ml/security_common';
+import { COMMON_REQUEST_HEADERS } from '../../../../functional/services/ml/common_api';
+
+export default ({ getService }: FtrProviderContext) => {
+  const ml = getService('ml');
+  const spacesService = getService('spaces');
+  const supertest = getService('supertestWithoutAuth');
+
+  const jobIdSpace1 = 'fq_single_space1';
+  const datafeedIdSpace1 = `datafeed-${jobIdSpace1}`;
+  const idSpace1 = 'space1';
+
+  async function getDatafeedById(
+    datafeedId: string | undefined,
+    expectedStatusCode: number,
+    space?: string
+  ) {
+    const { body } = await supertest
+      .get(`${space ? `/s/${space}` : ''}/api/ml/datafeeds${datafeedId ? `/${datafeedId}` : ''}`)
+      .auth(
+        USER.ML_VIEWER_ALL_SPACES,
+        ml.securityCommon.getPasswordForUser(USER.ML_VIEWER_ALL_SPACES)
+      )
+      .set(COMMON_REQUEST_HEADERS)
+      .expect(expectedStatusCode);
+
+    return body;
+  }
+
+  describe('update datafeeds with spaces', () => {
+    before(async () => {
+      await spacesService.create({ id: idSpace1, name: 'space_one', disabledFeatures: [] });
+
+      const jobConfig = ml.commonConfig.getADFqSingleMetricJobConfig(jobIdSpace1);
+      await ml.api.createAnomalyDetectionJob(jobConfig, idSpace1);
+      const datafeedConfig = ml.commonConfig.getADFqDatafeedConfig(jobIdSpace1);
+      await ml.api.createDatafeed(datafeedConfig, idSpace1);
+
+      await ml.testResources.setKibanaTimeZoneToUTC();
+    });
+
+    after(async () => {
+      await spacesService.delete(idSpace1);
+      await ml.api.cleanMlIndices();
+      await ml.testResources.cleanMLSavedObjects();
+    });
+
+    it('should update datafeed with correct space', async () => {
+      const newQueryDelay = '80000ms';
+
+      await ml.api.updateDatafeed(
```

Comment:
This test is validating the update API, so we should not use it from `ml.api` here because this sends the requests with superuser privileges. The general idea is that `ml.api` contains everything we need for test setup / teardown or supportive steps like waiting for states, fetching information (if it's not the API we're currently testing), while the tested APIs are defined in the test file where we could run the request with a non-superuser.
My suggestion is to move the code from `ml.api.updateDatafeed` to this file (similar to the `getDatafeedById` above) and move `getDatafeedById` to `ml.api` as we're not testing this API here but just using it to validate the success of the update API call.
This would then also allow us to add additional two tests to validate that only the ml_poweruser can update a datafeed and ml_viewer / ml_unauthorized can not.

Reply:
Updated in https://github.com/elastic/kibana/pull/124540/commits/97bc58c3e410d213f543232a4454e216a673296d
Moved `updateDatafeed` to test file, removed version from `ml.api`
Added test for updating datafeed in space it should not exist in.
Added test for updating with viewer user.